common/src/main/java/io/pravega/common/util/SortedDeque.java

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.base.Preconditions;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Deque-like data structure that accepts elements in sorted order and provides optimized operations based on that.
+ * Note: This class was inspired by java.util.ArrayDeque, however that class does not allow subclassing outside of its package.
+ */
+@NotThreadSafe
+@SuppressWarnings("unchecked")
+public class SortedDeque<E extends SortedIndex.IndexEntry> {
+    //region Members
+
+    private static final int MIN_INITIAL_CAPACITY = 8;
+    private transient SortedIndex.IndexEntry[] elements;
```

Comment:
I copied the base code from ArrayDeque, and thus these fields too. You're right, this is not needed.